### PR TITLE
DEP: stats.spearmanr: deprecate existing behavior with 2D inputs

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Part of gh-9307
gh-14651 

#### What does this implement/fix?
The behavior of `stats.spearmanr` with 2D input is inconsistent with all other stats reduction functions as described [here](https://github.com/scipy/scipy/issues/9307#issuecomment-1032070343). This PR deprecates the existing behavior, which will be replaced with more consistent behavior in SciPy 1.11.0.

#### Additional information
Already discussed [here](https://github.com/scipy/scipy/issues/9307#issuecomment-1116871761). [Email](https://mail.python.org/archives/list/scipy-dev@python.org/thread/XUEEXNZU6HFGCKXXWHCXA4DEE2JCDQUA/) sent to mailing list 5/6/2022.

### CODE REVIEW

File: scipy/stats/_stats_py.py [hunk 1]
```diff
@@ -4674,11 +4685,15 @@ def spearmanr(a, b=None, axis=0, nan_policy='propagate',
...
         b, _ = _chk_asarray(b, axis)
+        if b.ndim > 1:
+            warnings.warn(dep_msg, DeprecationWarning)
         if axisout == 0:
             a = np.column_stack((a, b))
         else:
             a = np.row_stack((a, b))
 
+
+
```
Comment (mdhaber) [inline scipy/stats/_stats_py.py:4696]:
```suggestion
```

**Comment by josef-pkt** [comment]
you cannot do this!

correlation functions like np.corrcoeff are mainly used as unary functions.
I almost never use them as function with two data arrays

```
spearmanr(np.random.rand(10, 3))[0].shape
(3, 3)
```

**Comment by josef-pkt** [comment]
I had an "ancient" Issue for whether a corr(x, y) should only return the cross correlation part instead of the full concatenated matrix.
where `corr` is any of the correlation functions

corr/cov   are matrix functions in general, i.e. should be vectorized as matrix functions

File: scipy/stats/_stats_py.py [hunk 2]
```diff
@@ -4662,7 +4662,18 @@ def spearmanr(a, b=None, axis=0, nan_policy='propagate',
...
+               "as of SciPy 1.9.0. In SciPy 1.11.0, the behavior will "
+               "change to be consistent with that of other reduction "
+               "functions, such as `f_oneway`. For example, the existing "
+               "behavior of `spearmanr(a)` with 2D `a` will be possible "
+               "with `spearmanr(a[:, np.newaxis, :], a[:, :, np.newaxis])`"
+               )
+
     a, axisout = _chk_asarray(a, axis)
+    if a.ndim > 1:
+        warnings.warn(dep_msg, DeprecationWarning)
```
Comment (j-bowhay) [inline scipy/stats/_stats_py.py:4675]:
```suggestion
        warnings.warn(dep_msg, DeprecationWarning, stacklevel=2)
```

File: scipy/stats/_stats_py.py [hunk 3]
```diff
@@ -4674,11 +4685,15 @@ def spearmanr(a, b=None, axis=0, nan_policy='propagate',
         # Concatenate a and b, so that we now only have to handle the case
         # of a 2-D `a`.
         b, _ = _chk_asarray(b, axis)
+        if b.ndim > 1:
+            warnings.warn(dep_msg, DeprecationWarning)
```
Comment (j-bowhay) [inline scipy/stats/_stats_py.py:4689]:
```suggestion
            warnings.warn(dep_msg, DeprecationWarning, stacklevel=2)
```

File: scipy/stats/_stats_py.py (hunk 2 above)
Comment (mdhaber) [reply scipy/stats/_stats_py.py:4675]: Thanks! I never remember!

**Comment by mdhaber** [comment]
@josef-pkt is this another way of stating your argument?

Correlation functions are not always though of as paired-sample statistics of univariate samples (like `ttest_rel`). Instead, you typically think of them as single-sample statistics of multivariate/vector-valued samples (like `np.cov`)?

**Comment by josef-pkt** [comment]
yes,
e.g. https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.corr.html

but the x, y both 1-dim case can be treated as special case to only return 1 scalar corr.

**Comment by mdhaber** [comment]
K. I'll leave this and other correlation functions alone for now, but I think there are still some decisions to be made before we add an `axis` argument to the other correlation functions.

**Pull request closed without merging**